src/include/parallel/task_executor.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2025, Sirius Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "task_scheduler.hpp"
+
+#include "duckdb/common/vector.hpp"
+
+#include <atomic>
+#include <thread>
+
+namespace sirius {
+namespace parallel {
+
+/**
+ * Interface for a thread pool used by different concrete executors like `GPUPipelineExecutor`, can be
+ * extended to support various kinds of tasks and scheduling policies.
+ */
+class ITaskExecutor {
+public:
+  struct WorkerThread {
+    explicit WorkerThread(duckdb::unique_ptr<std::thread> thread)
+      : internal_thread_(std::move(thread)) {}
+
+	  duckdb::unique_ptr<std::thread> internal_thread_;
+  };
+
+  struct Config {
+    int num_threads = std::thread::hardware_concurrency();
```

Comment:
the number of threads for task execution should correspond to the number of cudastreams we want scheduling work against the gpu. The default should be something more like 2-5 probably.